ksql-engine/src/main/java/io/confluent/ksql/function/udaf/topkdistinct/TopkDistinctKudaf.java

```diff
@@ -23,33 +23,47 @@
 import java.lang.reflect.Array;
 import java.util.Arrays;
 import java.util.Collections;
+import java.util.Comparator;
 import java.util.List;
 import java.util.Map;
 
 import io.confluent.ksql.function.KsqlAggregateFunction;
-import io.confluent.ksql.function.udaf.topk.TopkKudaf;
 import io.confluent.ksql.parser.tree.Expression;
 import io.confluent.ksql.util.ArrayUtil;
 import io.confluent.ksql.util.KsqlException;
 
-public class TopkDistinctKudaf<T> extends KsqlAggregateFunction<T, T[]> {
+public class TopkDistinctKudaf<T extends Comparable<? super T>>
+    extends KsqlAggregateFunction<T, T[]> {
 
-  private Integer tkVal;
-  private T[] tempTopkArray;
-  private Class<T> ttClass;
+  private final int tkVal;
+  private final Class<T> ttClass;
+  private final Comparator<T> comparator;
 
-  TopkDistinctKudaf(Integer argIndexInValue,
-                    Integer tkVal,
+  @SuppressWarnings("unchecked")
+  TopkDistinctKudaf(int argIndexInValue,
+                    int tkVal,
                     Class<T> ttClass) {
     super(argIndexInValue,
         () -> (T[]) Array.newInstance(ttClass, tkVal),
```

Comment:
not sure why this one wasn't also changed to not use `Array.newInstance(..)` ...

Reply:
See above ;)